include/wil/coroutine.h

```diff
@@ -0,0 +1,791 @@
+        auto as_handle() noexcept
+        {
+            return __WI_COROUTINE_NAMESPACE::coroutine_handle<Promise>::from_promise(*as_promise());
+        }
+
+        auto get_return_object() noexcept
+        {
+            // let the compiler construct the task / com_task from the promise.
+            return as_promise();
+        }
+
+        void destroy()
+        {
+            as_handle().destroy();
+        }
+
+        // The client lost interest in the coroutine, either because they are discarding
+        // the result without awaiting (risky!), or because they have finished awaiting.
+        // Discarding the result without awaiting is risky because any exception in the coroutine
+        // will be unobserved and result in a crash. If you want to disallow it, then
+        // raise an exception if waiting == running_ptr.
+        void abandon()
+        {
+            auto waiting = m_waiting.exchange(abandoned_ptr(), std::memory_order_acq_rel);
+            if (waiting != running_ptr()) destroy();
+        }
+
+        __WI_COROUTINE_NAMESPACE::suspend_never initial_suspend() noexcept
+        {
+            return {};
+        }
+
+        template<typename...Args>
+        void set_value(Args&&... args)
+        {
+            m_holder.set_value(wistd::forward<Args>(args)...);
+        }
+
+        void unhandled_exception() noexcept
+        {
+            m_holder.unhandled_exception();
+        }
+
+        void resume_waiting_coroutine(void* waiting) const
+        {
+            if (m_resumer)
+            {
+                m_resumer(waiting);
+            }
+            else
+            {
+                __WI_COROUTINE_NAMESPACE::coroutine_handle<>::from_address(waiting).resume();
+            }
+        }
+
+        auto final_suspend() noexcept
+        {
+            struct awaiter : __WI_COROUTINE_NAMESPACE::suspend_always
+            {
+                promise_base& self;
+                void await_suspend(__WI_COROUTINE_NAMESPACE::coroutine_handle<>) const noexcept
+                {
+                    // Need acquire so we can read from m_resumer.
+                    // Need release so that the results are published in the case that nobody
+                    // is awaiting right now, so that the eventual awaiter (possibly on another thread)
+                    // can read the results.
+                    auto waiting = self.m_waiting.exchange(completed_ptr(), std::memory_order_acq_rel);
+                    if (waiting == abandoned_ptr())
+                    {
+                        self.destroy();
+                    }
+                    else if (waiting != running_ptr())
+                    {
+                        WI_ASSERT(waiting != completed_ptr());
+                        self.resume_waiting_coroutine(waiting);
+                    }
+                };
+            };
+            return awaiter{ {}, *this };
+        }
+
+        // The remaining methods are used by the awaiters.

```

Comment:
These "client" awaiter functions are a great facility for creating custom ways for the task to produce its value. We needed a different implementation of `get()` in TTD (we still support Windows 7), for which we needed direct access to these.

We request the addition of public thin thunks to all three functions in the task_base class. Details of our findings are provided offline.